### PR TITLE
refactor!: restructure configuration to take options bag

### PR DESCRIPTION
**Pull request by aaronccasanova**
Per the discussion in #45 this PR restructures the current options API where each option is configured in three separate list and instead allows options to be configured in a single object.

The goal being to make the API more intuitive for configuring options (e.g. `short`, `withValue`, and `multiples`) while creating a path forward for introducing more configuration options in the future (e.g. `default`).

Before:
```js
parseArgs(process.argv.slice(2), {
  withValue: ['foo'],
  multiples: ['foo'],
  shorts: {
    f: 'foo',
  },
})
```

After:
```js
parseArgs({
  args: process.argv.slice(2),
  options: {
    foo: {
      short: 'f',
      type: 'string', // defaults to 'boolean'
      multiple: true,
      // default: 'bar' - Future iteration
    }
  },
})
```

While the above mentioned discussion left off with a tentative interest in moving forward with the suggestion, I wanted to at least make an official PR with the proposed changes and have an implementation we could reference.

Note: All tests are passing and I've introduced a few new validation helpers to check the new structure

TODO: Assuming we decide to move forward with this approach
- [x] Update README examples

### CODE REVIEW

**Comment by bcoe** [comment]
I'm at least partially sold on this approach, especially if it would help us reach consensus on some other features .. such as my thinking that we could figure out a better API surface for `strict` if we configure as an options bag.

**Comment by bcoe** [comment]
@aaronccasanova I'd be interested in seeing the updated README if you have the time.

One open question for me, if an argument is a flag, would you just set `withValue: false`? is this confusing (@ljharb, @shadowspawn). I like that we could support `strict` with this approach, without adding an additional configuration -- I just wonder if `withValue` is the right variable name.

**Comment by ljharb** [comment]
It is a bit confusing; I’d prefer an enum of “value” or “flag” if that were possible.

**Comment by aaronccasanova** [comment]
> @aaronccasanova I'd be interested in seeing the updated README if you have the time.

Will do 👍 

> if an argument is a flag, would you just set withValue: false? is this confusing

I agree this is a bit confusing. I was assuming users would omit `withValue` and leverage the zero config behavior where options without `withValue` are `flags`.

One alternative solution is to replace `withValue` with `type` (e.g. `type: string | number | boolean`). Not only does this allow users to explicitly define the option type, but provides more meaningful metadata to the parser. For example, If option was set to `type: 'number'` the parser could confidently capture negative numbers (`-100`) instead of expanding them as a short options group (`-1 -0 -0`) as mentioned [here](https://github.com/pkgjs/parseargs/pull/62#issuecomment-1035127531).

**Comment by bcoe** [comment]
> One alternative solution is to replace withValue with type (e.g. type: string | number | boolean)

I don't hate that idea, but for MVP think starting with `string` and `boolean` would be best, in the future we could decide whether we should add additional types, or leave this to implementers.

**Comment by aaronccasanova** [comment]
> for MVP think starting with string and boolean would be best

Agreed, I went ahead and replaced the `withValue` option with `type`, added validation checks for `string|boolean`, and updated the tests: https://github.com/pkgjs/parseargs/pull/63/commits/082bec5a5a667ee4d1d7725e4b8fcfa2199d0b4c

**Comment by aaronccasanova** [comment]
@bcoe Updated the README to reflect the most recent implementation: https://github.com/pkgjs/parseargs/pull/63/commits/09090081a42c5eea90a338e6cd1a2d2bfe1ec978

File: README.md
Comment (aaronccasanova) [inline README.md:74]: Any interest in representing this as a type alias? This is an extremely biased opinion, but this is much more readable to me. Understanding, this could be a barrier for folks not familiar with `TypeScript` conventions (such as: `?` optional object properties)
e.g.
```ts
/**
  * `parseArgs` configuration options
  */
type config = {
  /**
   * Array of argument strings
   * @default process.mainArgs
   */
  argv?: string[];
  /**
   * A collection of configuration objects for each `argv`
   */
  options?: {
    [longName: string]: {
      /**
        * Type of `argv`
        * @default 'boolean'
        */
      type?: 'string' | 'boolean'; 
      /**
       * If true, when appearing multiple times in `argv`, will be concatenated into an Array
       * @default false
       */
      multiples?: boolean;
      /**
       * An alias to an `argv`; When appearing multiples times in `argv`; Respects the `multiples` configuration
       */
      short?: string; 
    };
  };
  /**
    * A Boolean on whether or not to throw an error when unknown args are encountered
    */
  strict?: boolean;
}



```

**Comment by bnb** [comment]
Added the Merge into Node.js milestone, not because this needs to be merged by then but because IMO we should definitely figure this out - merge or close - before that IMO.

File: README.md
Comment (aaronccasanova) [reply README.md:74]: Scratch this.. I'm not seeing this convention used anywhere in existing Node documentation.

**Comment by isaacs** [comment]
I like this a lot. It also leaves the door open for things like `description`, `options`, `conflictsWith`, `validate: () => Boolean`, etc.  And it's much more declarative/readable.

File: README.md
Comment (shadowspawn) [inline README.md:78]: > A collection of configuration objects for each `argv`

This mention of `argv` does not seem correct. Perhaps:

> An object describing the known options to look for in `argv`

Comment (shadowspawn) [reply README.md:78]: > `options` keys are the long names of the `argv`

Likewise, use of `argv` does not seem correct. Perhaps:

>  `options` keys are the long names of the known options

File: README.md
Comment (shadowspawn) [inline README.md:79]: > Type of `argv`

Suggest: Type of known option

(I am adding "known" partly because there can be unknown options,  but more to give some context for the many uses of "option". If it adds more doubt in your mind than helps, feel free to omit!)

File: README.md
Comment (shadowspawn) [inline README.md:80]: (This usage of `argv` does make sense.)

File: README.md
Comment (shadowspawn) [inline README.md:81]: 1) 

> An alias to an `argv`

Suggest: a single character alias for an option

2) Suggest drop the mention of multiples entirely. (It was present in the original, so feel free to treat as out-of-scope for this PR!)

File: README.md
Comment (shadowspawn) [reply README.md:80]: Bonus: the current implementation intentionally returns even a single usage of the option as an array. The wording could be updated to reflect this, or you could consider it out-of-scope for this PR. 😄

File: README.md
Comment (shadowspawn) [reply README.md:79]: > defaults to `'boolean'`

👍

File: README.md
Comment (aaronccasanova) [reply README.md]: Used both suggestions! Thanks for the feedback: https://github.com/pkgjs/parseargs/pull/63/commits/e44c46c758cd8882ae7ad520323f514eac40691b

File: README.md
Comment (aaronccasanova) [reply README.md]: I like it 👍 https://github.com/pkgjs/parseargs/pull/63/commits/e44c46c758cd8882ae7ad520323f514eac40691b

File: README.md
Comment (aaronccasanova) [reply README.md]: > returns even a single usage of the option as an array

Good point, updated `multiples` to reflect that, thoughts?
https://github.com/pkgjs/parseargs/pull/63/commits/e44c46c758cd8882ae7ad520323f514eac40691b#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R80

File: README.md
Comment (aaronccasanova) [reply README.md]: > Suggest: a single character alias for an option

Great suggestion! I agree it's best to be explicit here as the current implementation always expands short option groups and thus only single character shorts are valid 👍https://github.com/pkgjs/parseargs/pull/63/commits/e44c46c758cd8882ae7ad520323f514eac40691b#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R81

File: README.md
Comment (shadowspawn) [reply README.md]: Better! 👍

**Review by bcoe — APPROVED** [review]
👍 I think this is an improvement to the API:

1. it solves the problem of `strict`, by giving us `type` (which we can use to define flag options explicitly).
2. it simplifies configuration so that we don't need to define options for an argument in multiple places.

**Comment by bcoe** [comment]
Proposing we land this early in the week, if we don't see any arguments against here or in #45.

It feels like doing so would sure up some important remaining work, and could actually put us on track for being in Node 18.

**Comment by shadowspawn** [comment]
Non-blocking suggestion: property could be `args` rather than `argv`?

```
const result = parseArgs({ args: process.mainArgs, strict:false });
```

**Comment by ljharb** [comment]
but `process.argv` is a thing, and `process.mainArgs` isn't?

**Comment by shadowspawn** [comment]
Yes, I did consider real `process.argv` vs proposed `process.mainArgs`. (And wondering about asking where `mainArgs` was headed!) But that wasn't the only reason I suggested it.

Factors in favour of `args`:

1) Part of the point of `parseArgs` is to get away from the arcane conventions of `argv`.

2) If I have some custom arguments to parse in my own code, I would rather call them `args` than `argv`.

```js
const result = parseArgs({ args: myArgs, strict:false });
```

```js
const args = getTestArgs();
const result = parseArgs({ args, strict:false });
```

3) The routine is called `parseArgs` not `parseArgv`

4) If I did want to parse `process.argv`, I don't need to specify a property at all.

**Comment by ljharb** [comment]
I’m not sure how arcane it is; it’s still the terminology in the environment the arguments are coming from :-0

**Comment by shadowspawn** [comment]
Yeah, nah, I am still liking `args`. Some scenarios.

1) I want the arguments from the Node.js environment my application was launched with.

a) leave it up to `parseArgs` (normal)

```
const result = parseArgs({ strict:false });
```

b) specify it myself (perhaps after preprocessing, say).  Examples with both naming.

```
const result = parseArgs({ argv: process.argv, strict:false }); // [sic] wrong, argv is not argv
const result = parseArgs({ args: process.argv.slice(2), strict:false }); // get the args from argv
```

2) I have my own arguments.

```
const args = getTestArgs();
const result = parseArgs({ args, strict:false });
```

**Comment by shadowspawn** [comment]
Note that @aaronccasanova made exactly the `argv: process.argv` mistake in his description of usage! (Sorry Aaron, but useful example. 😄 )

https://github.com/pkgjs/parseargs/pull/63#issue-1127954190

**Comment by bakkot** [comment]
Since `process.argv` is not the intended/default value, +1 to not naming this `argv`.

**Comment by aaronccasanova** [comment]
> Note that @aaronccasanova made exactly the `argv: process.argv` mistake in his description of usage! (Sorry Aaron, but useful example. 😄)

Lol no worries! I actually intentionally used `process.argv` instead of `process.argv.slice(2)` to keep the example concise and focused on the options API. In regards to `args` vs `argv`, my [initial implementation used `args`](https://github.com/pkgjs/parseargs/pull/63/commits/697493f23f7be34e33cce5151f0bfeafcb68b499#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R82) and I updated it to `argv` to reduce the number of changes introduced in this PR. Although, I also prefer `args` for two reasons:
- The default/input args will likely always be processed and not representative of the standard `argv` shape
- The utility is called `parseArgs` and thus `args` is a more intuitive name for the inputs (imo)

File: README.md
Comment (shadowspawn) [inline README.md:83]: > `options` keys are the long names of the known options

Does this mean I can not add configuration for an option which only has a short form?

(I do not consider this a blocker, and can be solved. Just making sure people are aware of the new limitation.)

**Comment by shadowspawn** [comment]
Weak suggestion: `multiple` rather than `multiples`. (The plural in the original was because it was an array of options.)

File: index.js
Comment (shadowspawn) [inline index.js:59]: We are thrashing `option` name! I don't think I like `options` being a collection of options which is indexed by a variable called `option`. 

Perhaps index it using parameter named `optionKey` or `optionName` to get the "option". 

(Or `optionsConfig` to match `optionConfig` in following line.)

File: index.js
Comment (shadowspawn) [inline index.js:88]: Same comment about using `option` to index into `options` to get an `optionConfig`. (Disclaimer: I have not tried this at scale to check it helps, so feel free to question the concept and I'll give it a try!.)

File: index.js
Comment (shadowspawn) [reply index.js:59]: Or perhaps `longOption` (instead of `optionKey` or `optionName`) to align with the potential short/long naming.

Comment (shadowspawn) [reply index.js:59]: (I don't mind the "option" from `options` being called `optionConfig`.)

Poke me for clarification if that was all to ambiguous!

File: index.js
Comment (shadowspawn) [inline index.js:96]: Can/should we validate this is a single character?

Comment (ljharb) [reply index.js:96]: Yes.

File: validators.js
Comment (shadowspawn) [inline validators.js:20]: Do you need to use an Array primordial for `includes` here for robustness in face of funky prototypes?

**Review by shadowspawn — CHANGES_REQUESTED** [review]
I added a number of suggestions. The one I care about most are:
- change property from `argv` to `args`
- validate single character for `short`
- primordial needed for `include` (?)

Negotiable.

File: validators.js
Comment (shadowspawn) [reply validators.js:20]:
```suggestion
  if (!ArrayPrototypeIncludes(union, value)) {
```

File: index.js
Comment (ljharb) [inline index.js:88]: for..of is never robust, because of Symbol.iterator; this should use `ArrayPrototypeForEach(ObjectEntries(options), ([option, optionConfig]) => {` instead.

File: index.js
Comment (ljharb) [inline index.js:145]:
```suggestion
        const foundLong = ArrayPrototypeFind(ObjectEntries(options), ([, optionConfig]) => optionConfig.short === arg);
        arg = foundLong?.[0] ?? StringPrototypeCharAt(arg, 1);
```

File: validators.js
Comment (ljharb) [inline validators.js:21]:
```suggestion
    throw new ERR_INVALID_ARG_TYPE(name, `[${StringPrototypeJoin(union, '|')}]`, value);
```

**Comment by shadowspawn** [comment]
Light comment: in terms of shim correctness, I am guessing the new validation helpers need to be outside `validators.js` which is a core file?

**Comment by shadowspawn** [comment]
I'll go through and edit in some `non-blocking` explanations in some of my suggestions, as most do not affect client, are an existing shortcoming that can be remedied in a small PR, etc.

I am keen on changing `argv` to `args` as it is widespread and client facing, and have submitted a PR for that.

File: README.md
Comment (aaronccasanova) [reply README.md:83]: > Does this mean I can not add configuration for an option which only has a short form?

No you can. Just pass a single character as the long name: e.g.
```js
// node long-short.js -f bar
const {options} = parseArgs({
  options: {
    f: {
      type: 'string'
    }
  }
})
options // { f: 'bar' }
```

File: index.js
Comment (aaronccasanova) [reply index.js:59]: That was clear 👍 I like `longOption` and `optionKey`. I'm leaning towards `optionKey` to align with the utilities you're introducing in #68. However, `longOption` is more semantic and is more inline with our efforts to improve the readability of the source. Thoughts?

File: index.js
Comment (aaronccasanova) [reply index.js:88]: I'm all for it! https://github.com/pkgjs/parseargs/pull/63#discussion_r816256277

File: index.js
Comment (shadowspawn) [reply index.js:59]: I think `longOption` might reduce chance for confusion between short and long forms and which can be used as key, and reflects new structure. (And the confusion will be less of a problem after this refactor anyway.)

File: README.md
Comment (shadowspawn) [reply README.md]: I think that will add configured `--f` but not `-f`. In particular, `-f` will not be found by this code:

```
        arg = StringPrototypeCharAt(arg, 1); // short
        for (const [option, optionConfig] of ObjectEntries(options)) {
          if (optionConfig.short === arg) {
            arg = option; // now long!
            break;
          }
```

File: index.js
Comment (aaronccasanova) [reply index.js]: https://github.com/pkgjs/parseargs/pull/63/commits/b0b71c43ca10677ac9048cb9f7b1eff2c1d35987

File: README.md
Comment (shadowspawn) [reply README.md]: I misunderstood the implementation, I'll add a new comment and hide the above.

Comment (shadowspawn) [reply README.md]: A quirk of the current implementation is that a one character long option will be recognised as both a long option (e.g. `--f`) and a short option (e.g. `-f`). This is an existing quirk, and not introduced by this PR.

File: index.js
Comment (shadowspawn) [inline index.js:99]: Cosmetic, optional suggestion:
```suggestion
        throw new Error(`options.${option}.short must be a single character, got '${short}'`);
```

File: index.js
Comment (shadowspawn) [reply index.js:59]: This is a non-blocking comment on the naming. (And assuming we land this PR first, I can revisit the naming when I redo the refactor of the code per #68.)

**Review by shadowspawn — APPROVED** [review]
Only action items on my list are the two `for` loops noted by @ljharb 
Approving assuming either fix those now, or open an issue to do later. 😄

File: index.js
Comment (aaronccasanova) [reply index.js]: 👌 https://github.com/pkgjs/parseargs/pull/63/commits/ecf1fb09a855ccef0f4a4eaba410edb9e90610ed

File: index.js
Comment (aaronccasanova) [reply index.js]: Thanks! https://github.com/pkgjs/parseargs/pull/63/commits/ecf1fb09a855ccef0f4a4eaba410edb9e90610ed

File: validators.js
Comment (aaronccasanova) [reply validators.js]: 👌 https://github.com/pkgjs/parseargs/pull/63/commits/e0d0c3356eb61154ad826cf6fd85b49596589778

File: README.md
Comment (aaronccasanova) [reply README.md]: I see no issue with this unless we think there are cases where authors do NOT want to allow shorts for a given option. Is this a scenario you've encountered maintaining Commander? Or foresee other problems with this quirk?

File: index.js
Comment (aaronccasanova) [reply index.js]: `longOption` it is!! Honestly, this update totally improved readability (imo): https://github.com/pkgjs/parseargs/pull/63/commits/50d4f4d605c9e0da1f201b482961d039e40571ee

File: index.js
Comment (aaronccasanova) [reply index.js]: 👌 https://github.com/pkgjs/parseargs/pull/63/commits/b41fd9bee71f7e6aa6dc63560a739eec3e2ca0b0

**Comment by aaronccasanova** [comment]
Thanks everyone for the feedback and helping improve the robustness of the implementation!! I think the PR is in a really good spot and has laid a great foundation for future iterations. Here are some notable changes that have recently been introduced since the initial reviews:
* The `argv` parser config has been renamed to `args`
* The `multiples` option config has been renamed to `multiple`
* Internal usages of the name `option` have been updated to `longOption` and `shortOption` respectively
* Added validation to ensure the `short` option config is a single character

File: README.md
Comment (shadowspawn) [reply README.md]: Commander started out with dual short and long option names (needed to define both). Support for long-only options added in 2011. Support for short-only options added in 2020 (and more for completeness than because of demand).

File: index.js
Comment (shadowspawn) [reply index.js]: Sweet. 🎉

File: index.js
Comment (shadowspawn) [reply index.js]: Woop!

**Review by shadowspawn — APPROVED** [review]
LGTM. Thanks for the enthusiasm and work @aaronccasanova

File: index.js
Comment (ljharb) [inline index.js:150]: nit:
```suggestion
        const [longOption] = ArrayPrototypeFind(
```

**Review by ljharb — APPROVED** [review]

**Pull request merged**
